vyper/venom/analysis/mem_ssa.py

```diff
@@ -0,0 +1,388 @@
+        self.inst_to_def: Dict[IRInstruction, MemoryDef] = {}
+        self.inst_to_use: Dict[IRInstruction, MemoryUse] = {}
+
+    def analyze(self):
+        # Request required analyses
+        self.cfg = self.analyses_cache.request_analysis(CFGAnalysis)
+        self.dom = self.analyses_cache.request_analysis(DominatorTreeAnalysis)
+        self.alias = self.analyses_cache.request_analysis(MemoryAliasAnalysis)
+
+        # Build initial memory SSA form
+        self._build_memory_ssa()
+
+        # Clean up unnecessary phi nodes
+        self._remove_redundant_phis()
+
+    def mark_location_volatile(self, loc: MemoryLocation) -> MemoryLocation:
+        volatile_loc = self.alias.mark_volatile(loc)
+
+        for bb in self.memory_defs:
+            for mem_def in self.memory_defs[bb]:
+                if self.alias.may_alias(mem_def.loc, loc):
+                    assert mem_def.loc is not None
+                    new_loc = MemoryLocation(
+                        offset=mem_def.loc.offset, size=mem_def.loc.size, is_volatile=True
+                    )
+                    mem_def.loc = new_loc
+
+        return volatile_loc
+
+    def get_memory_def(self, inst: IRInstruction) -> Optional[MemoryDef]:
+        return self.inst_to_def.get(inst)
+
+    def get_memory_use(self, inst: IRInstruction) -> Optional[MemoryUse]:
+        return self.inst_to_use.get(inst)
+
+    def _build_memory_ssa(self):
+        """Build the memory SSA form for the function"""
+        # Initialize entry block with liveOnEntry
+        entry_block = self.dom.entry_block
+        self.current_def[entry_block] = self.live_on_entry
+
+        for bb in self.cfg.dfs_pre_walk:
+            self._process_block_definitions(bb)
+
+        # Second pass: insert phi nodes where needed
+        self._insert_phi_nodes()
+
+        # Third pass: connect uses to their reaching definitions
+        self._connect_uses_to_defs()
+
+    def _process_block_definitions(self, block: IRBasicBlock):
+        """Process memory definitions and uses in a basic block"""
+        effect_type = Effects.STORAGE if self.location_type == "storage" else Effects.MEMORY
+        for inst in block.instructions:
+            # Check for memory reads
+            if effect_type in inst.get_read_effects():
+                mem_use = MemoryUse(self.next_id, inst)
+                self.next_id += 1
+                self.memory_uses.setdefault(block, []).append(mem_use)
+                self.inst_to_use[inst] = mem_use
+
+            # Check for memory writes
+            if effect_type in inst.get_write_effects():
+                mem_def = MemoryDef(self.next_id, inst)
+                self.next_id += 1
+
+                mem_def.reaching_def = self._get_reaching_def_for_def(block, mem_def)
+
+                self.memory_defs.setdefault(block, []).append(mem_def)
+                self.current_def[block] = mem_def
+                self.inst_to_def[inst] = mem_def
+
+    def _insert_phi_nodes(self):
```

Comment:
since this doesn't actually insert physical phi instructions, can we call it like `compute_phi_nodes` or similar?

Reply:
I don't know... feels like insert to me as we are inserting those into the memory graph (even though the graph is now separate from the other data structures, compared to our "instruction based" variable ssa phi)

Reply:
the relatedly named `_add_phi_nodes` function in MakeSSA does add physical instructions, so it's a bit confusing
https://github.com/vyperlang/vyper/blob/b396d124b05e45905d62b0e109ed363e5570b2b9/vyper/venom/passes/make_ssa.py#L34-L37

at least add a comment saying that it's adding the nodes to the graph, not instructions to the basic block, otherwise it seems like the function body does not match the name